// @cord.network/modules/lib/cjs/stream/Stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyAgainstDocument = exports.isIStream = exports.fromDocument = exports.verifyDataStructure = void 0;
const tslib_1 = require("tslib");
const utils_1 = require("@cord.network/utils");
const Did = tslib_1.__importStar(require("@cord.network/did"));
const Document = tslib_1.__importStar(require("../document/index.js"));
/**
 *  Checks whether the input meets all the required criteria of an [[IStream]] object.
 *  Throws on invalid input.
 *
 * @param input The potentially only partial [[IStream]].
 *
 */
function verifyDataStructure(input) {
    if (!input.identifier) {
        throw new utils_1.SDKErrors.IdentifierMissingError();
    }
    utils_1.DataUtils.validateId(input.identifier, 'Stream Identiifier');
    if (!input.schema) {
        throw new utils_1.SDKErrors.SchemaMissingError();
    }
    utils_1.DataUtils.validateId(input.schema, 'Schema Identifier');
    if (!input.streamHash) {
        throw new utils_1.SDKErrors.StreamHashMissingError();
    }
    utils_1.DataUtils.verifyIsHex(input.streamHash, 256);
    if (!input.issuer) {
        throw new utils_1.SDKErrors.IssuerMissingError();
    }
    Did.validateUri(input.issuer, 'Did');
    if (typeof input.revoked !== 'boolean') {
        throw new utils_1.SDKErrors.RevokedTypeError();
    }
}
exports.verifyDataStructure = verifyDataStructure;
/**
 * Builds a new instance of an [[Stream]], from a complete set of input required for an stream.
 *
 * @param document - The base request for stream.
 * @returns A new [[Stream]] object.
 *
 */
function fromDocument(document) {
    const stream = {
        identifier: document.identifier,
        streamHash: document.documentHash,
        issuer: document.content.issuer,
        schema: document.content.schemaId,
        authorization: document.authorization,
        registry: document.registry,
        revoked: false,
    };
    verifyDataStructure(stream);
    return stream;
}
exports.fromDocument = fromDocument;
/**
 * Custom Type Guard to determine input being of type IStream using the StreamUtils errorCheck.
 *
 * @param input The potentially only partial IStream.
 * @returns Boolean whether input is of type IStream.
 */
function isIStream(input) {
    try {
        verifyDataStructure(input);
    }
    catch (error) {
        return false;
    }
    return true;
}
exports.isIStream = isIStream;
/**
 * Verifies whether the data of the given attestation matches the one from the corresponding credential. It is valid if:
 * * the [[Credential]] object has valid data (see [[Credential.verifyDataIntegrity]]);
 * and
 * * the hash of the [[Credential]] object, and the hash of the [[Stream]].
 *
 * @param stream - The stream to verify.
 * @param document - The document to verify against.
 */
function verifyAgainstDocument(stream, document) {
    const schemaMismatch = document.content.schemaId !== stream.schema;
    const documentMismatch = document.documentHash !== stream.streamHash;
    if (documentMismatch || schemaMismatch) {
        throw new utils_1.SDKErrors.CredentialUnverifiableError(`Some attributes of the stream diverge from the credential: ${[
            'schemaId',
            'streamHash',
        ]
            .filter((_, i) => [schemaMismatch, documentMismatch][i])
            .join(', ')}`);
    }
    Document.verifyDataIntegrity(document);
}
exports.verifyAgainstDocument = verifyAgainstDocument;
